pin/set.go

```diff
@@ -0,0 +1,327 @@
+	}
+	if err := writeHdr(n, hdr); err != nil {
+		return nil, err
+	}
+	hdrLen := len(n.Data)
+
+	if estimatedLen < maxItems {
+		// it'll probably fit
+		for i := 0; i < maxItems; i++ {
+			k, data, ok := iter()
+			if !ok {
+				// all done
+				break
+			}
+			n.Links = append(n.Links, &merkledag.Link{Hash: k.ToMultihash()})
+			n.Data = append(n.Data, data...)
+		}
+		// sort by hash, also swap item Data
+		s := sortByHash{
+			links: n.Links[defaultFanout:],
+			data:  n.Data[hdrLen:],
+		}
+		sort.Stable(s)
+	}
+
+	// wasteful but simple
+	type item struct {
+		k    util.Key
+		data []byte
+	}
+	hashed := make(map[uint32][]item)
+	for {
+		k, data, ok := iter()
+		if !ok {
+			break
+		}
+		h := hash(seed, k)
+		hashed[h] = append(hashed[h], item{k, data})
+	}
+	for h, items := range hashed {
+		childIter := func() (k util.Key, data []byte, ok bool) {
+			if len(items) == 0 {
+				return "", nil, false
+			}
+			first := items[0]
+			items = items[1:]
+			return first.k, first.data, true
+		}
+		child, err := storeItems(ctx, dag, uint64(len(items)), childIter, internalKeys)
+		if err != nil {
+			return nil, err
+		}
+		size, err := child.Size()
+		if err != nil {
+			return nil, err
+		}
+		childKey, err := dag.Add(child)
+		if err != nil {
+			return nil, err
+		}
+		internalKeys(childKey)
+		l := &merkledag.Link{
+			Name: "",
+			Hash: childKey.ToMultihash(),
+			Size: size,
+			Node: child,
+		}
+		n.Links[int(h%defaultFanout)] = l
+	}
+	return n, nil
+}
+
+func readHdr(n *merkledag.Node) (*pb.Set, []byte, error) {
+	hdrLenRaw, consumed := binary.Uvarint(n.Data)
+	if consumed <= 0 {
+		return nil, nil, errors.New("invalid Set header length")
+	}
+	buf := n.Data[consumed:]
+	if hdrLenRaw > uint64(len(buf)) {
+		return nil, nil, errors.New("impossibly large Set header length")
+	}
+	// as hdrLenRaw was <= an int, we now know it fits in an int
+	hdrLen := int(hdrLenRaw)
+	var hdr pb.Set
+	if err := proto.Unmarshal(buf[:hdrLen], &hdr); err != nil {
+		return nil, nil, err
+	}
+	buf = buf[hdrLen:]
+
+	if v := hdr.GetVersion(); v != 1 {
+		return nil, nil, fmt.Errorf("unsupported Set version: %d", v)
+	}
+	if uint64(hdr.GetFanout()) > uint64(len(n.Links)) {
+		return nil, nil, errors.New("impossibly large Fanout")
+	}
+	return &hdr, buf, nil
+}
+
+func writeHdr(n *merkledag.Node, hdr *pb.Set) error {
+	hdrData, err := proto.Marshal(hdr)
+	if err != nil {
+		return err
+	}
+	n.Data = make([]byte, binary.MaxVarintLen64, binary.MaxVarintLen64+len(hdrData))
+	written := binary.PutUvarint(n.Data, uint64(len(hdrData)))
+	n.Data = n.Data[:written]
+	n.Data = append(n.Data, hdrData...)
+	return nil
+}
+
+type walkerFunc func(buf []byte, idx int, link *merkledag.Link) error
+
+func walkItems(ctx context.Context, dag merkledag.DAGService, n *merkledag.Node, fn walkerFunc, children keyObserver) error {
+	hdr, buf, err := readHdr(n)
+	if err != nil {
+		return err
+	}
+	// readHdr guarantees fanout is a safe value
+	fanout := int(hdr.GetFanout())
```

Comment:
`func (m *Set) GetFanout() uint32`, later fanout is used for slice index; it's needed.